R/mwrite.R

```diff
@@ -0,0 +1,415 @@
+    }
+  }
+  
+  if("PROB" %in% names(code)) {
+    l$prob <- code$PROB
+    code$PROB <- NULL
+  }
+  
+  clob <- c("PARAM", "INPUT", "THETA", "CMT", "INIT", "OMEGA", "SIGMA", 
+            "NMEXT", "NMXML", "VCMT", "SET", "CAPTURE")
+  for(block in clob) {
+    while(block %in% names(code)) {
+      code[[block]] <- NULL
+    }
+  }
+  
+  # Need special handling here in case compartments are declared 
+  # as a part of PKMODEL
+  if("PKMODEL" %in% names(code)) {
+    pk <- tolist(code$PKMODEL)
+    parsed <- do.call(PKMODEL, pk)
+    pk$ncmt <- parsed$n
+    pk$cmt <- NULL
+    code$PKMODEL <- tocode(pk)
+  }
+  # Put block names back into code
+  code <- Map(code, names(code), f = function(text, name) {
+    c(glue("${name}"), text, " ")
+  })
+  l$code <- unlist(code, use.names = FALSE)
+  l
+}
+
+#' Write model code to yaml  format
+#' 
+#' Model code is written to a readable, transport format. This transport format
+#' can be useful for (1) breaking connection to NONMEM modeling outputs that 
+#' are imported by `$NMXML` or `$NMEXT` and (2) saving model updates (e.g., 
+#' an updated parameter list). Models can be read back using [mread_yaml()] or 
+#' converted to mrgsolve cpp format with [yaml_to_cpp()].
+#' 
+#' @param x a model object. 
+#' @param file output file name; if non-character (e.g., `NULL`), no output 
+#' will be written to file. 
+#' @param digits precision to use when writing outputs. 
+#' 
+#' @details
+#' Parameters and omega and sigma matrices that were imported via `$NMXML`
+#' or `$NMEXT` will be written into the yaml file and the NONMEM import blocks 
+#' will be dropped. This allows the user to load a model based on a NONMEM run 
+#' without having a connection to that output (e.g., `root.xml` or `root.ext`). 
+#' Given that the connection to the NONMEM modeling outputs is broken when 
+#' writing to yaml, any update to the NONMEM run will only be propagated to 
+#' the yaml file when `mwrite_yaml()` is run again. 
+#' 
+#' The yaml file does not currently have the ability to track 
+#' other external dependencies, such as user-defined header files or other 
+#' code that might be sourced in by the user when the model is loaded via 
+#' [mread()]. NONMEM xml and ext files imported by `$NMXML` or `$NMEXT` are 
+#' the _only_ external dependencies that are accounted for in the yaml 
+#' transport file. 
+#' 
+#' @return 
+#' A list containing data that was written out to the yaml file, with added 
+#' item `file`, is returned invisibly. 
+#' 
+#' @examples
+#' mod <- house()
+#' 
+#' temp1 <- tempfile(fileext = ".yaml")
+#' 
+#' x <- mwrite_yaml(mod, temp1)
+#' 
+#' readLines(temp1)
+#' 
+#' @seealso [mread_yaml()], [yaml_to_cpp()]
+#' 
+#' @md
+#' @export
+mwrite_yaml <- function(x, file, digits = 8) {
+  require_yaml()
+  l <- mwrite_model_to_list(x)
+  l$format <- "yaml"
+  out <- yaml::as.yaml(l, precision = digits)
+  if(is.character(file)) {
+    writeLines(con = file, out)  
+  }
```

Comment:
This `writeLines` call adds an additional newline compared to what `write_yaml` would produce. I think it'd be good to align those.

You could do that by tweaking the `sep` value:

```R
tfile <- tempfile()
x <- list(one = 1, two = 2)

yaml::write_yaml(x, tfile)
readChar(tfile, file.size(tfile))
#> [1] "one: 1.0\ntwo: 2.0\n"

writeLines(yaml::as.yaml(x), tfile)
readChar(tfile, file.size(tfile))
#> [1] "one: 1.0\ntwo: 2.0\n\n"

writeLines(yaml::as.yaml(x), tfile, sep = "")
readChar(tfile, file.size(tfile))
#> [1] "one: 1.0\ntwo: 2.0\n"
```

Or perhaps just switch to using `write_yaml`?

```R
if(is.character(file)) {
  l$format <- "yaml"
  yaml::write_yaml(l, file = file, precision = digits)
}
```